[jit] add initial support for sparse tensors

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23842 [jit] support grad and data attribute for tensor in script
* **#23841 [jit] add initial support for sparse tensors**
* #23639 [jit] support dict augment assign in script

Summary:
This PR add the initial support for sparse tensors in TorchScript, as
part of the effort to support whole training loop and optimizers. It
allows the sparse tensor to be passed in and out of the TorchScript, but
a lot of the optimization pathways havenot been thorougly tested with
sparse tensors, and we need to disable/bail out the CUDA fusion when one
of the input tensor is sparse tensor in our fusion group and use the
fallback JIT interpreter (because our CUDA code gen does not handle the
sparse tensor yet).

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23841

Differential Revision: [D16683560](https://our.internmc.facebook.com/intern/diff/D16683560)

wanchaol committed:
[jit] add initial support for sparse tensors

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

wanchaol committed:
Update on "[jit] add initial support for sparse tensors"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

wanchaol committed:
Update on "[jit] add initial support for sparse tensors"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D16683560](https://our.internmc.facebook.com/intern/diff/D16683560)

zdevito commented on torch/csrc/jit/pybind_utils.h:
```diff
@@ -328,7 +325,8 @@ inline IValue toIValue(
       if (var.is_sparse()) {
-        AT_ERROR("sparse tensors not supported");
+        AT_WARN(
+            "Sparse tensors in TorchScript is experimental and likely to break");
```
-> "Using sparse tensors in TorchScript is experimental. Many optimization pathways have not been thoroughly tested with sparse tensors. Please include the fact that the network is running sparse tensors in any bug reports submitted."

zdevito reviewed (approved):
Please update the summary of the diff to provide context.

wanchaol committed:
Update on "[jit] add initial support for sparse tensors"


Summary:
This PR add the initial support for sparse tensors in TorchScript, as
part of the effort to support whole training loop and optimizers. It
allows the sparse tensor to be passed in and out of the TorchScript, but
a lot of the optimization pathways havenot been thorougly tested with
sparse tensors, and we need to disable/bail out the CUDA fusion when one
of the input tensor is sparse tensor in our fusion group and use the
fallback JIT interpreter (because our CUDA code gen does not handle the
sparse tensor yet).

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23841

Differential Revision: [D16683560](https://our.internmc.facebook.com/intern/diff/D16683560)

wanchaol committed:
rebase and address comments on "[jit] add initial support for sparse tensors"


Summary:
This PR add the initial support for sparse tensors in TorchScript, as
part of the effort to support whole training loop and optimizers. It
allows the sparse tensor to be passed in and out of the TorchScript, but
a lot of the optimization pathways havenot been thorougly tested with
sparse tensors, and we need to disable/bail out the CUDA fusion when one
of the input tensor is sparse tensor in our fusion group and use the
fallback JIT interpreter (because our CUDA code gen does not handle the
sparse tensor yet).

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23841

Differential Revision: [D16683560](https://our.internmc.facebook.com/intern/diff/D16683560)
